src/libraries/System.Text.Json/src/System/Text/Json/Document/JsonElement.cs

```diff
@@ -1242,92 +1277,127 @@ internal static bool DeepEquals(JsonElement left, JsonElement right)
                     return true;
 
                 case JsonValueKind.Number:
-                    return JsonHelpers.AreEqualJsonNumbers(left.GetRawValue().Span, right.GetRawValue().Span);
+                    return JsonHelpers.AreEqualJsonNumbers(element1.GetRawValue().Span, element2.GetRawValue().Span);
 
                 case JsonValueKind.String:
-                    if (right.ValueIsEscaped)
+                    if (element2.ValueIsEscaped)
                     {
-                        if (left.ValueIsEscaped)
+                        if (element1.ValueIsEscaped)
                         {
                             // Both values are escaped, force an allocation to unescape the RHS.
-                            return left.ValueEquals(right.GetString());
+                            return element1.ValueEquals(element2.GetString());
                         }
 
                         // Swap values so that unescaping is handled by the LHS.
-                        (left, right) = (right, left);
+                        (element1, element2) = (element2, element1);
                     }
 
-                    return left.ValueEquals(right.ValueSpan);
+                    return element1.ValueEquals(element2.ValueSpan);
```

Comment:
Why is it that `element1.ValueEquals(element2.GetString())` needs to be used if element2 is escaped? `element1.ValueEquals(element2.ValueSpan)` doesn't produce the correct answer if both element1 and element2 are escaped?

Reply:
`ValueEquals` always assumes that the RHS doesn't contain escaping.